Add support for system text scale factor.

gspencergoog opened the pull request:
This fixes #10524, adding support to the engine and the android platform code for responding to system settings for font scaling factor.

I added these messages to the system channel, as a "systemSettings" type with a single double member called "textScaleFactor".

It triggers the hooks and sets the internal value on the Window object.

goderbauer commented on shell/common/engine.cc:
```diff
@@ -44,6 +44,7 @@ constexpr char kAssetChannel[] = "flutter/assets";
 constexpr char kLifecycleChannel[] = "flutter/lifecycle";
 constexpr char kNavigationChannel[] = "flutter/navigation";
 constexpr char kLocalizationChannel[] = "flutter/localization";
+constexpr char kSystemChannel[] = "flutter/system";
```
Since the memory message and the text scale message seem to get processed at two different places it might make sense to have a separate channel for the text scale after all instead of trying to separate the messages on the same system channel?

Do engine folks have an opinion on this? @cbracken @jason-simmons @Hixie

gspencergoog replied:
I think if I added another channel, I would make it more general than just text scale.  Maybe "flutter/settings" or something.  Is there any place where the different channels are documented (e.g. what they are supposed to be for, what their scope should be, naming conventions, etc)?

Hixie replied:
I don't think we have such documentation yet, but we should. Please file a bug to make sure we write such docs and add them to the API docs.

gspencergoog replied:
OK, Done.  Filed [flutter/flutter#12196](https://github.com/flutter/flutter/issues/12196)

Changed the channel to "flutter/settings".

gspencergoog commented:
So is this good to go?  I want to commit it so I can start rolling the engine to get the flutter side in.

Hixie commented on lib/ui/testing_hooks.dart:
```diff
@@ -0,0 +1,19 @@
+  /// Do not use for non-testing purposes: catastrophic problems may occur.
+  static void updateTextScaleFactor(double textScaleFactor) =>
+      _updateTextScaleFactor(textScaleFactor);
+}
```
We should definitely not expose an API we don't want people to use.

Hixie replied:
Any API we expose _will_ be used, and as soon as its used we can't break it, regardless of what our documentation says (especially if we hide it with `@nodoc`).

gspencergoog replied:
Well, in principle, I agree, but there isn't any other way that I know of to simulate these events.  I discussed this with Jason before implementing it, and he seemed to think this was the best compromise.

gspencergoog replied:
Out of curiosity: why are we _especially_ bound to @nodoc APIs?  Seems like just the opposite to me.

Hixie replied:
We talked about this at lunch but for anyone watching:

It's probably not literally true that `@nodoc` is more likely to make people depend on it. There's a combination of factors that could push that way though:
- if we say `@nodoc`, then anything we put in the docs will be missed by anyone who decides to use it.
- people especially like to use secret APIs because it makes them feel cool and hip.
- if we say it's not supported then that lowers our expectation that people will use it, but it doesn't lower the actual probability that people will use it, so the end result is that relative to our expectation, people are more likely to end up depending on it.

At the end of the day though Flutter's convention is to not have "secret" APIs, we either have public documented supported APIs, or we don't expose the API at all, and in particular, our next layer should always use the previous layer's public API so that people can replace any layer and do the same thing we did (this is true at the macro level, e.g. "rendering" vs "dart:ui", as well as the micro level, e.g. Icon using IconTheme).

We want to avoid magic.

Hixie commented:
I recommend testing the engine side of this using the part-of hack in <https://github.com/flutter/engine/blob/master/testing/dart/window_hooks_integration_test.dart>. To test whether this actually results in the MediaQuery changing you need probably need a devicelab test where you actually have the system-level font size change or something like that.

Hixie commented:
cc @jason-simmons for a more detailed review of the C++ side.

goderbauer commented on lib/ui/testing_hooks.dart:
```diff
@@ -0,0 +1,19 @@
+/// Do not use for non-testing purposes, or catastrophic problems may (and
+/// probably will) occur.  These APIs are not part of the Flutter API, and are
+/// subject to change.  Do not depend upon their existence.
+class TestingHooks {
```
Exposing this seems bad... is there no other way?

gspencergoog commented:
Thanks for the integration hooks pointer.  I'll look into using that instead.

gspencergoog commented:
I signed it!

gspencergoog commented:
I signed it!

gspencergoog commented:
OK, well, googlebot aside, I think this is ready, so PTAL.  I've added the integration hooks test, and removed the TestingHooks class.  Hixie and I talked about other ways to test, and decided that it would be really cool to have a method channel that is just for testing (and only enabled in debug builds) that would allow of simulation of engine events.  I'll do that in another PR, though.

goderbauer commented on testing/dart/window_hooks_integration_test.dart:
```diff
@@ -125,65 +140,101 @@ void main() {
       runZoned(() {
         innerZone = Zone.current;
-        window.onSemanticsAction = (_, __) {
+        window.onSemanticsAction = (int value, __) {
```
nit: reduce the number of underscores for second arg?

gspencergoog replied:
Done.

goderbauer commented on testing/dart/window_hooks_integration_test.dart:
```diff
@@ -55,55 +57,68 @@ void main() {
       expect(runZone, isNotNull);
       expect(runZone, same(innerZone));
+      expect(devicePixelRatio, isNotNull);
+      expect(devicePixelRatio, equals(0.1234));
```
Isn't equals(1234) implying that it cannot be null?

goderbauer replied:
see also a couple of times further down.

gspencergoog replied:
Yeah, I guess you're right.  I add those because when the test fails because it's null, I thought it would be nicer to have it highlight that it was null, but I guess it's pretty obvious for the equals case.

goderbauer commented on testing/dart/window_hooks_integration_test.dart:
```diff
@@ -125,65 +140,101 @@ void main() {
       runZoned(() {
         innerZone = Zone.current;
-        window.onPlatformMessage = (_, __, ___) {
+        window.onPlatformMessage = (String value, __, ___) {
```
same

Hixie commented on testing/dart/window_hooks_integration_test.dart:
```diff
@@ -125,65 +137,99 @@ void main() {
       });
 
-      _dispatchPlatformMessage(null, null, null);
+      _dispatchPlatformMessage("testName", null, null);
```
nit: use single quotes unless there's a reason to use double quotes

gspencergoog replied:
Done.  (It's my years of C++ fingers that need retraining...)

Hixie commented on testing/dart/window_hooks_integration_test.dart:
```diff
@@ -125,65 +137,99 @@ void main() {
       runZoned(() {
         innerZone = Zone.current;
-        window.onSemanticsAction = (_, __) {
+        window.onSemanticsAction = (int value, _) {
```
also, not a big deal, but as a general rule we should always include types (even for unused arguments), that way when someone is looking at the code it's more self-documenting, easier to copy/paste, etc. Don't worry about adding the type (and maybe explicit variable name) here unless you happy to be on this line some time, though.

gspencergoog replied:
I think for tests that's less of an issue.  I think it adds to readability of the test to be able to just visually skip over unused arguments (because they are so often irrelevant).  I agree that for framework code this is reversed, and that they should always have types.
